nvme.c

```diff
@@ -5317,6 +5398,629 @@ static int show_registers(int argc, char **argv, struct command *cmd, struct plu
+		if (err)
+			return err;
+		fabrics = true;
+	}
+
+	if (!get_register_offset(bar, fabrics, &cfg, flags)) {
+		nvme_show_error("offset required param");
+		err = -EINVAL;
+	}
+
+	if (fabrics)
+		free(bar);
+	else
+		munmap(bar, getpagesize());
+
+	return err;
+}
+
+static void nvme_set_register(void *bar, int offset, uint64_t value, bool mmio32)
+{
+	if (nvme_is_64bit_reg(offset))
+		mmio_write64(bar + offset, value, mmio32);
+	else
+		mmio_write32(bar + offset, value);
+
+	printf("set-register: %02x (%s), value: %"PRIx64"\n", offset,
+	       nvme_register_to_string(offset), value);
+}
+
+static void set_register_value(void *bar, struct argconfig_commandline_options *opts,
+			       struct set_reg_config *cfg, int offset)
+{
+	switch (offset) {
+	case NVME_REG_INTMS:
+		if (argconfig_parse_seen(opts, "intms"))
+			nvme_set_register(bar, offset, cfg->intms, cfg->mmio32);
+		break;
+	case NVME_REG_INTMC:
+		if (argconfig_parse_seen(opts, "intmc"))
+			nvme_set_register(bar, offset, cfg->intmc, cfg->mmio32);
+		break;
+	case NVME_REG_CC:
+		if (argconfig_parse_seen(opts, "cc"))
+			nvme_set_register(bar, offset, cfg->cc, cfg->mmio32);
+		break;
+	case NVME_REG_CSTS:
+		if (argconfig_parse_seen(opts, "csts"))
+			nvme_set_register(bar, offset, cfg->csts, cfg->mmio32);
+		break;
+	case NVME_REG_NSSR:
+		if (argconfig_parse_seen(opts, "nssr"))
+			nvme_set_register(bar, offset, cfg->nssr, cfg->mmio32);
+		break;
+	case NVME_REG_AQA:
+		if (argconfig_parse_seen(opts, "aqa"))
+			nvme_set_register(bar, offset, cfg->aqa, cfg->mmio32);
+		break;
+	case NVME_REG_ASQ:
+		if (argconfig_parse_seen(opts, "asq"))
+			nvme_set_register(bar, offset, cfg->asq, cfg->mmio32);
+		break;
+	case NVME_REG_ACQ:
+		if (argconfig_parse_seen(opts, "acq"))
+			nvme_set_register(bar, offset, cfg->acq, cfg->mmio32);
+		break;
+	case NVME_REG_BPRSEL:
+		if (argconfig_parse_seen(opts, "bprsel"))
+			nvme_set_register(bar, offset, cfg->bprsel, cfg->mmio32);
+		break;
+	case NVME_REG_CMBMSC:
+		if (argconfig_parse_seen(opts, "cmbmsc"))
+			nvme_set_register(bar, offset, cfg->cmbmsc, cfg->mmio32);
+		break;
+	case NVME_REG_NSSD:
+		if (argconfig_parse_seen(opts, "nssd"))
+			nvme_set_register(bar, offset, cfg->nssd, cfg->mmio32);
+		break;
+	case NVME_REG_PMRCTL:
+		if (argconfig_parse_seen(opts, "pmrctl"))
+			nvme_set_register(bar, offset, cfg->pmrctl, cfg->mmio32);
+		break;
+	case NVME_REG_PMRMSCL:
+		if (argconfig_parse_seen(opts, "pmrmscl"))
+			nvme_set_register(bar, offset, cfg->pmrmscl, cfg->mmio32);
+		break;
+	case NVME_REG_PMRMSCU:
+		if (argconfig_parse_seen(opts, "pmrmscu"))
+			nvme_set_register(bar, offset, cfg->pmrmscu, cfg->mmio32);
+		break;
+	default:
+		break;
+	}
+}
+
+static int set_register_check(struct argconfig_commandline_options *opts,
+			      struct set_reg_config *cfg, int offset)
+{
+	bool flag = false;
+
+	switch (offset) {
+	case NVME_REG_INTMS:
+		if (argconfig_parse_seen(opts, "intms"))
```

Comment:
Seems like it should not be necessary to go through the flags twice. Just have a single function that checks the flags and parses out the offset and value to set?

Reply:
Sorry for the complicated code implementation but this function is to check if the named argument register is duplicated with the offset argument register or not. But the function `set_register_value()` checks to set the named argument register by using `argconfig_parse_seen()`. I will do consider to refactor the code to be more simple later.

Reply:
I guess I don't really see the need for this check. If `--offset` is specified along with a specific offset argument for the same offset, what's the harm? I guess one of them will overwrite the other? Doesn't seem worth all this checking code to me.

Reply:
This check is for example the `--offset` specified as `0x14` CC register offset then `--value` value used to set but if the `--cc` also specified the register set to the option value as below so duplicated the register settings and needed to prevent to set.
```sh
nvme set-reg /dev/nvme0 --offset=0x14 --value=0x460000 --cc=0x350001
```

Reply:
I am also sitting on the fence if all the code to handle `--offset` and `--regname` is really necessary. From the user POV it's obviously nice to have both incl error handling.